### DISCUSSION TITLE
Allow using boa in an async context

### DISCUSSION THREAD

**Discussion by CodeDoctorDE** [boa-dev/boa#4001] [General]
Hello,

I'm currently trying to add boa to flutter. Based on the feedback of the bridge author (https://github.com/fzyzcjy/flutter_rust_bridge/discussions/2257) Context needs to be sendable (implement Send).
Can you add it? Do you have a better idea how I could integrate it? I'm currently using a `Arc<Mutex<Context>>` for this.

**Comment by jedel1043** [comment]
Converted to discussion because this is more of a question thing.

**Comment by jedel1043** [comment]
In general, a `Context` cannot be `Send` because the ECMAScript spec is pretty clear that execution contexts are single-threaded. This fact is used by Boa with some thread local data for certain things like garbage collecting, which automatically makes `Context` not `Send` since it keeps track of objects that can only be accessed by the thread local GC.

I think the most common workaround for this is to have a thread local `Context`, which should allow accessing it globally, with the caveat that any new thread would create its own `Context` and `Realm` to execute things on.

**Reply by thautwarm** [reply]
As I'm one who also uses both flutter_rust_bridge and boajs, I believe this might not satisfy the use case.
Boa is for scripting, and I don't really need multi-thread processing here.
Hence, a workaround might be creating a worker thread to run boajs, and async stuffs can communicate with the worker thread using messages that are `Send`.

**Reply by jedel1043** [reply]
Yeah, if you don't need multithreading and if you don't run scripts often enough to cause contention between threads, a dedicated thread for the engine would also work
